### DISCUSSION TITLE
Simplifying CellProfiler's Export modules

### DISCUSSION THREAD

**Discussion by bethac07** [CellProfiler/CellProfiler#4714] [General]
I've started a branch (`export_module`), but adding a post here describing the high-level strategy stuff for feedback.

Basically, the Export modules live in the space of "similar enough that there's a lot of duplication, dissimilar enough that it's often hard to figure out why in Spreadsheets you get X but in Databases you get Y". This is not a great situation.

We've been talking about doing a merge and cleanup for at least as long as I've been in the group, so let's make CP5 the version where we get it done.

Vision:
- Deprecate EtD and EtS
- Create new "ExportData" module
  - The actual-actual work of writing to X output format will get extracted out to new `exporter/format.py` files per format, probably in the utilities directory - this keeps things semi-organized, and makes it easy to add or deprecate individual formats as we like.
  - The ExportData module handles all the settings choices about what we're saving and where
    - Unless actually technically IMPOSSIBLE to implement for a particular backend, most options now become generic across output formats. No more "you can pick your measurements when writing to a spreadsheet but not a database" or "you can have a checkbox export the Object relationships info in a database but not in a spreadsheet"
- CPA Properties/Workspace stuff go to new "CreateCellProfilerAnalystFiles" module
- We will try as we're doing this to resolve/address the 20+ open issues for the Export modules

Anticipated challenges:
- Probably at least some of the things that are output-format-specific are so for a reason. We'll find out which as we go, presumably.
- Upgrade path - this is going to be definitely not-trivial in terms of upgrading existing pipelines with regards to both merging data exports AND handling properties file creation if it already existed 
- We will probably want to institute some sort of checker that when CreateCellProfilerAnalystFiles is in the pipeline, that it has ExportData and with a compatible format - unclear to me without digging into the internals more if this is birds or national parks

**Comment by jenna-tomkinson** [comment]
Hi Beth!

Is this initial info still relevant or in progress?

We're interested in finding out what will become of exports in CellProfiler!

Thanks!

Best,
Jenna, Dave (@d33bs), Mike (@MikeLippincott), Erik (@axiomcura)

**Reply by bethac07** [reply]
Hi Jenna,

Yes, though progress has been slow due to lots of competing priorities - we still want this to be in CP5 but have no additional implementation details at this time (it will probably not be worked on at least until CellProfiler-library is complete). We of course always are happy to have contributions!
